keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,993 @@
+As a batch workload user (e.g., MPI/TensorFlow/Spark), I want all tasks of a 
+Job to be either started altogether; otherwise, don’t start any of them.
+
+#### Story 2
+As a service workload user, I want to use PodGroup along with other scheduling
+directives to compose a customized scheduling requirement - like, I want a
+Deployment to run in the same topology (using PodAffinity) at the same time 
+(using PodGroup); otherwise, don’t start any of them.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Most batch jobs consist of multiple (usually two) roles of tasks. For example, a
+typical Tensorflow job has some parameter servers and workers. Each role
+requires a quorum to be ready to run. With that being considered, the following
+API designs all employ a 2-tier specification (Subsets) inside PodGroup. 
+
+The PodGroup API will be defined as a top-level API (in group scheduling.k8s.io, like PriorityClass) 
+
+```go
+// group scheduling.k8s.io
+type PodGroup struct {
+	Spec   PodGroupSpec
+	Status PodGroupStatus
+}
+
+// PodGroupSpec is a description of a PodGroupSpec.
+type PodGroupSpec struct {
+	// Subsets consist of various kinds of pod sets.
+	// A PodGroup is schedulable if all subsets can be scheduled.
+	Subsets []Subset
+
+	// ScheduleTimeoutSeconds defines the timeout threshold to abort
+	// an in-progress PodGroup-level scheduling attempt.
+	ScheduleTimeoutSeconds *int32
+}
+
+// Subset represents a collection of pods with the same role.
+type Subset struct {
+	// Name is used to distinguish pods with different roles in the
+	// same pod group. Optional if there is only one subset in the 
+	// pod group.
+	Name string
+
+	// MinMember specifies the minimum number of pods required for a 
+	// subset to be operational.
+	MinMember int32
+}
+```
+
+In order to associate a pod with its target PodGroup, we need to add a field call `PodGroup` in the podSpec to declare which PodGroup and Subset the pod belongs to.
+
+```go
+// PodSpec is a description of a pod.
+type PodSpec struct {
+  ....
+  PodGroup PodGroupSpec
+	....
```

Comment:
Could you format it properly? Same as L349~352.

Reply:
there's prior art for this... it's similar to storageclass, which was added to storage.k8s.io, went through alpha → beta → ga, and was referenced from v1.PersistentVolumeClaim

Reply:
btw: IIRC storageclass was hosted as a CRD in alpha, and moved to in-tree in beta. For PodGroup, it's better to reside in-tree since day 1.